src/ck-core/ckarray.C

```diff
@@ -1034,21 +1045,57 @@ void CkArray::insertElement(CkMarshalledMessage&& m, const CkArrayIndex& idx,
 }
 
 /// This method is called by ck.C or the user to add an element.
-bool CkArray::insertElement(CkArrayMessage* me, const CkArrayIndex& idx,
+bool CkArray::insertElement(CkArrayMessage* m, const CkArrayIndex& idx,
                             int listenerData[CK_ARRAYLISTENER_MAXLEN])
 {
   CK_MAGICNUMBER_CHECK
   int onPe;
+  // Element's sibling already lives somewhere else, so insert there instead.
   if (locMgr->isRemote(idx, &onPe))
-  { /* element's sibling lives somewhere else, so insert there */
-    thisProxy[onPe].insertElement(me, idx, listenerData);
+  {
+    thisProxy[onPe].insertElement(m, idx, listenerData);
     return false;
   }
-  int ctorIdx = me->array_ep();
+  int ctorIdx = m->array_ep();
   int chareType = _entryTable[ctorIdx]->chareIdx;
-  ArrayElement* elt = allocate(chareType, me, false, listenerData);
-  if (!locMgr->addElement(thisgroup, idx, elt, ctorIdx, (void*)me))
+  ArrayElement* elt = allocate(chareType, m, false, listenerData);
+
+  // Get the ID for this element
+  CmiUInt8 id = locMgr->getNewObjectID(idx);
```

Comment:
Instead of (physically) inlining all of this logic, maintaining some degree of separation may be valuable. Particularly, separating the logic for registration and delivery into a helper function like `ckBeforeConstruction` or `ckRegisterElement` would help to keep `insertElement` concise.

Reply:
I've made some changes along these lines, although with entry method invocation being routed through CkLocRec, I could not fully remove that relationship yet. In the future though, I want CkLocRec either gone, or just a passive struct holding some small amount of information that is specific to a "location" rather than an "element".